src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject_Options.cs

```diff
@@ -29,21 +31,50 @@ protected void UpdateOptions()
         }
 
         /// <summary>
-        /// Sets the given command line arguments to be the last parsed command line arguments.
+        /// If the command line has changed from the last parsed command line, then it parses it and sets new command line arguments.
+        /// Subsequently, regardless of whether command line changed, it creates and sets new options using the last parsed command line arguments.
         /// </summary>
-        protected void SetArguments(CommandLineArguments commandLineArguments)
+        protected CommandLineArguments SetArgumentsAndUpdateOptions(string commandlineForOptions)
         {
-            _lastParsedCommandLineArguments = commandLineArguments;
+            SetArguments(commandlineForOptions);
+            UpdateOptions();
+            return _lastParsedCommandLineArguments;
         }
 
         /// <summary>
-        /// Sets the given command line arguments to be the last parsed command line arguments and
-        /// creates and sets new options using these command line arguments.
+        /// Resets the last parsed command line and updates options with the same command line.
         /// </summary>
-        protected void SetArgumentsAndUpdateOptions(CommandLineArguments commandLineArguments)
+        /// <remarks>
+        /// Use this method when options can go stale due to side effects, even though the command line is identical.
+        /// For example, changes to contents of a ruleset file needs to force update the options for the same command line.
+        /// </remarks>
+        protected CommandLineArguments ResetArgumentsAndUpdateOptions()
         {
-            SetArguments(commandLineArguments);
-            UpdateOptions();            
+            // Clear last parsed command line.
+            var savedLastParsedCompilerOptions = _lastParsedCompilerOptions;
+            _lastParsedCompilerOptions = null;
+            _lastParsedCommandLineArguments = null;
+
+            // Now set arguments and update options with the saved command line.
+            return SetArgumentsAndUpdateOptions(savedLastParsedCompilerOptions);
+        }
+
+        /// <summary>
+        /// If the command line has changed from the last parsed command line, then it parses it and sets new command line arguments.
+        /// </summary>
+        /// <param name="commandlineForOptions"></param>
+        /// <returns></returns>
+        protected CommandLineArguments SetArguments(string commandlineForOptions)
+        {
+            if (!string.Equals(_lastParsedCompilerOptions, commandlineForOptions, StringComparison.OrdinalIgnoreCase))
```

Comment:
I don't agree that this should be `OrdinalIgnoreCase`.  That makes it impossible to update the the options for case changes of output file paths, plus there are many options which are not actually paths.